protocol/consistencychecks.go

```diff
@@ -0,0 +1,328 @@
+	uname string, key []byte) ErrorCode {
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var str *m.SignedTreeRoot
+	var err error
+
+	switch requestType {
+	case RegistrationType:
+		str, err = cc.verifyRegistration(requestType, msg, uname, key)
+	case KeyLookupType:
+		str, err = cc.verifyKeyLookup(requestType, msg, uname, key)
+	case MonitoringType:
+	case KeyLookupInEpochType:
+	default:
+		panic("[coniks] Unknown request type")
+	}
+
+	// update the state
+	if err == PassedWithAProofOfAbsence ||
+		err == PassedWithAProofOfInclusion {
+		cc.SavedSTR = str
+	}
+
+	return err.(ErrorCode)
+}
+
+func (cc *ConsistencyChecks) verifyRegistration(requestType int,
+	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
+
+	df, ok := msg.DirectoryResponse.(*DirectoryProof)
+	if !ok {
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	ap := df.AP
+	str := df.STR
+
+	// 1. verify STR
+	if err := cc.verifySTR(cc.SavedSTR, str); err != nil {
+		return nil, err
+	}
+
+	// 2. verify Auth path
+	proofType, err := verifyAuthPath(uname, key, ap, str)
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. verify (response code, proof type) pair
+	switch {
+	case msg.Error == ErrorNameExisted:
+	case msg.Error == Success && proofType == proofOfAbsence:
+	default:
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	// 4. verify returned promise
+	if err := cc.verifyReturnedPromise(requestType,
+		msg.Error, df, uname, key); err != nil {
+		return nil, err
+	}
+
+	if proofType == proofOfAbsence {
+		return str, PassedWithAProofOfAbsence
+	}
+	return str, PassedWithAProofOfInclusion
+}
+
+func (cc *ConsistencyChecks) verifyKeyLookup(requestType int,
+	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
+
+	df, ok := msg.DirectoryResponse.(*DirectoryProof)
+	if !ok {
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	ap := df.AP
+	str := df.STR
+
+	// 1. verify STR
+	if err := cc.verifySTR(cc.SavedSTR, str); err != nil {
+		return nil, err
+	}
+
+	// 2. verify Auth path
+	proofType, err := verifyAuthPath(uname, key, ap, str)
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. verify (response code, proof type) pair
+	switch {
+	case msg.Error == ErrorNameNotFound && proofType == proofOfAbsence:
+	case msg.Error == Success:
+	default:
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	// determine which kind of TB's verification we should do
+	// based on the response code and proof type
+	if msg.Error == Success && proofType == proofOfAbsence {
+		// 4.1 verify returned promise
+		if err := cc.verifyReturnedPromise(requestType,
+			msg.Error, df, uname, key); err != nil {
+			return nil, err
+		}
+	} else {
+		// 4.2 verify fulfilled promise
+		if err := cc.verifyFulfilledPromise(uname, str, ap); err != nil {
+			return nil, err
+		}
+	}
+
+	if proofType == proofOfAbsence {
+		return str, PassedWithAProofOfAbsence
+	}
+	return str, PassedWithAProofOfInclusion
+}
```

Comment:
Hmm, I just refactored auth path verification in 41cc371674f0c22c484718cbb22d8543a2988f7e, so it would reduce some repeated checks of `proofType`. Then I can do `(proofType, error, reqType)` checks earlier. What do you think? I probably should extract the change in `proof.go` into separate PRs, but I would like to have your comment about this.
/cc @Liamsi.